src/float.rkt

```diff
@@ -71,8 +50,111 @@
 (define (bit-difference x y)
   (ulps->bits (+ 1 (abs (ulp-difference x y)))))
 
-(define (sample-float)
-  (floating-point-bytes->real (integer->integer-bytes (random-exp 32) 4 #f)))
+(define (ordinary-value? x)
+  (match x
+    [(? real?)
+     (not (or (infinite? x) (nan? x)))]
+    [(? complex?)
+     (and (ordinary-value? (real-part x)) (ordinary-value? (imag-part x)))]
+    [(? boolean?)
+     true]
+    [(? posit8?)
+     (not (set-member? (representation-special-values posit8) x))]
+    [(? posit16?)
+     (not (set-member? (representation-special-values posit16) x))]
+    [(? posit32?)
+     (not (set-member? (representation-special-values posit32) x))]
+    [_ true]))
+
+(module+ test
+  (check-true (ordinary-value? 2.5))
+  (check-false (ordinary-value? +nan.0))
+  (check-false (ordinary-value? -inf.f)))
+
+(define (=-or-nan? x1 x2)
+  (cond
+    [(and (number? x1) (number? x2))
+     (or (= x1 x2)
+         (and (nan? x1) (nan? x2)))]
+    [(and (posit8? x1) (posit8? x2))
+     (= ((representation-repr->ordinal posit8) x1)
+        ((representation-repr->ordinal posit8) x2))]
+    [(and (posit16? x1) (posit16? x2))
+     (= ((representation-repr->ordinal posit16) x1)
+        ((representation-repr->ordinal posit16) x2))]
+    [(and (posit32? x1) (posit32? x2))
+     (= ((representation-repr->ordinal posit32) x1)
+        ((representation-repr->ordinal posit32) x2))]
+    ;; TODO once we have real ->ordinal for quires, fix this
+    [(and (quire8? x1) (quire8? x2))
+     (= ((representation-repr->ordinal quire8) x1)
+        ((representation-repr->ordinal quire8) x2))]
+    [(and (quire16? x1) (quire16? x2))
+     (= ((representation-repr->ordinal quire16) x1)
+        ((representation-repr->ordinal quire16) x2))]
+    [(and (quire32? x1) (quire32? x2))
+     (= ((representation-repr->ordinal quire32) x1)
+        ((representation-repr->ordinal quire32) x2))]))
+
+(module+ test
+  (check-true (=-or-nan? 2.3 2.3))
+  (check-false (=-or-nan? 2.3 7.8))
+  (check-true (=-or-nan? +nan.0 -nan.f))
+  (check-false (=-or-nan? 2.3 +nan.f)))
+
+(define (</total x1 x2)
+  (cond
```

Comment:
Presumably we should rewrite this to interrogate the repr and do it in a repr-generic way, right?

Reply:
Right, I think this will be another change that's made a lot easier once we start passing through the precision though.